### DISCUSSION TITLE
Tell Larastan a query builder is a specific kind using interface?

### DISCUSSION THREAD

**Discussion by rudiedirkx** [larastan/larastan#1931] [Q&A]
I have a list of models that I want to query using a scope that they share (in name, not logic).

```php
$models = [
    User::class => [1, 2],
    Post::class => [3, 4],
    Tag::class => [5, 6],
];
```

and then I want to load those models:

```php
foreach ($models as $class => $ids) {
    $models[$class] = $class::query()->forTimeline()->find($ids);
}
```

Larastan magically thinks `$class::query()` is an `EloquentBuilder`, even though it doesn't know what `$class` is, but okay. But `forTimeline()` doesn't exist on `EloquentBuilder`, so I have to be more specific somehow. My models implement `TimelineModel`:

```php
interface TimelineModel {
	public function scopeForTimeline(Builder $query) : void;
}
```

so maybe I can use that?

```php
foreach ($models as $class => $ids) {
    /* @var class-string<TimelineModel> $class */
    $models[$class] = $class::query()->forTimeline()->find($ids);
}
```

No, that doesn't work, that's too complex, I admit. Add a fake static method on the interface?

```php
/**
 * @method static EloquentBuilder<self> forTimeline()
 */
```

Also no. The `$query` dumped type is `EloquentBuilder<TimelineModel>`, but I'm not sure that's even understandable for Larastan. And PHPStan complains anyway: 

> TimelineModel in generic type EloquentBuilder<TimelineModel> in PHPDoc tag @var for variable $query is not subtype of template type TModelClass of EloquentModel of class EloquentBuilder

because how would it know TimelineModel is always a Model..?

Maybe with a specific step in between?

```php
foreach ($models as $class => $ids) {
    /* @var EloquentBuilder<TimelineModel> $query */
    $query = $class::query();
    $models[$class] = $query->forTimeline()->find($ids);
}
```

Also no!

Only this works, but I don't want to repeat class names:

```php
foreach ($models as $class => $ids) {
    /* @var EloquentBuilder<User|Post|Tag> $query */
    $query = $class::query();
    $models[$class] = $query->forTimeline()->find($ids);
}
```

I hope the intention is clear. I want Larastan to understand what type $query is, but by interface, not class. Is that even possible? Does that even make sense...? 🤔 What would the return type even be...? Maybe my interface approach doesn't make sense... But I can't map `$models`/`$class` to specific class-strings, the process before that is too complicated, no way static analysis is going to understand that.

**Comment by calebdw** [comment]
You should take a look into custom query builders which helps with static analysis and removes a lot of the laravel magic

**Reply by rudiedirkx** [reply]
But it's not 1 query builder, it's 3, with 1 shared method. That's why I like the 1 scope. And it works perfectly, in reality/PHP, not in theory/Stan.

**Reply by calebdw** [reply]
You can apply interfaces to custom query builders as well.

However, if you insist on using the scope then you need to use an intersection type to let phpstan know that the class is a model **and** it implements the interface, something like this should work:

```php
foreach ($models as $class => $ids) {
    /* @var class-string<Model&TimelineModel> $class */
    $models[$class] = $class::query()->forTimeline()->find($ids);
}
// or
foreach ($models as $class => $ids) {
    /* @var EloquentBuilder<Model&TimelineModel> $query */
    $query = $class::query();
    $models[$class] = $query->forTimeline()->find($ids);
}
```

**Reply by rudiedirkx** [reply]
Larastan isn't quite that smart:

> Call to an undefined method EloquentBuilder<Model&TimelineModel>::forTimeline().

probably because `TimelineModel` has a `scopeForTimeline()`, not a `forTimeline()`. Too much Laravel magic. I'll just keep my specific classes: `EloquentBuilder<User|Post|Tag>`.

**Reply by calebdw** [reply]
This is fixed by https://github.com/larastan/larastan/pull/1979
